### DISCUSSION TITLE
CustomPlaylistServiceTest 관련 질문

### DISCUSSION THREAD

**Discussion by SeonghaeJo** [wafflestudio/seminar-2023#192] [Q&A] [answered]
안녕하세요.
CustomPlaylistServiceTest에서 "커스텀 플레이리스트 생성시, 자동 생성되는 제목은 내 플레이리스트 #${유저의 커스텀 플레이리스트 갯수 + 1}" 테스트를 수행할 때 문제가 생겨 질문 드립니다.

<img width="765" alt="스크린샷 2023-11-04 오전 2 57 15" src="https://github.com/wafflestudio/seminar-2023/assets/87258768/6a98b4da-1b76-431d-bf39-e1a4d305cfaa">

위 사진처럼 테스트 함수를 개별적으로 수행할 때는 해당 테스트 함수는 물론 CustomPlaylistServiceTest에 있는 모든 테스트 함수가 성공적으로 통과됩니다.
그런데 CustomPlaylistServiceTest에 있는 테스트를 한번에 수행하게 할 때나, `./gradlew test`를 수행할 때는 `커스텀 플레이리스트 생성시, 자동 생성되는 제목은 내 플레이리스트 #${유저의 커스텀 플레이리스트 갯수 + 1}`보다 `존재하지 않는 곡을 추가 시도시, SongFoundException이 발생한다`와 `커스텀 플레이리스트에 곡 추가시, songCnt가 올라가고 연결 테이블의 row가 생성된다`를 먼저 테스트하고, 그 테스트 과정에서 create를 2번 호출합니다. 이 때문에 아래 사진처럼 `커스텀 플레이리스트 생성시, 자동 생성되는 제목은 내 플레이리스트 #${유저의 커스텀 플레이리스트 갯수 + 1}` 테스트를 수행할 때 이미 커스텀 플레이리스트가 2개 생성되어 있어 "내 플레이리스트 #1"으로 하드코딩되어 있는 테스트에 실패하는 것 같습니다.

<img width="1302" alt="스크린샷 2023-11-04 오전 2 56 24" src="https://github.com/wafflestudio/seminar-2023/assets/87258768/87587474-45e9-4733-8b46-bfdd8a8b5d74">

"주의: @Transactional 사용하지 않기. (데이터 롤백, 레이지 로드 등의 이슈를 피할 수 있도록 구현)" 부분을 고려해서 테스트 함수에 @Transactional을 추가할 수는 없을 것 같은데, 어떻게 해결하면 좋을까요? 제 생각엔 앞선 2개의 테스트에서 create가 2번 호출되고 커밋된 후에 다른 테스트 함수를 수행할 때 데이터베이스를 롤백하지 않으니(@Transactional이 없으므로) 위 사진처럼 "내 플레이리스트 #3"이 나오는 게 맞아 질문드립니다.
감사합니다.

**Comment by PFCJeong** [comment, marked as answer]
안녕하세요~

요 부분 제가 명세를 잘못했는지 헷갈려하시는 분들이 계신 것 같네요ㅠ 죄송합니다.

과제에서 요구하는 것은, 해당 테스트에서 `@Transactional`을 사용하지 않고 테스트에서 생성된 데이터를 롤백하는 것입니다.

예전에 테스트 데이터 롤백 관련 질문이 들어와 관련해서 답변을 드린 것이 있는데, [참고](https://github.com/wafflestudio/seminar-2023-springboot-assignment/pull/42#pullrequestreview-1651843145)하시면 될 것 같습니다.

**Reply by SeonghaeJo** [reply]
이해했습니다! 감사합니다
